File: app/src/main/java/com/aquamorph/frcmanager/Activities/MainActivity.java
package com.aquamorph.frcmanager.activities;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.SharedPreferences.OnSharedPreferenceChangeListener;
import android.content.res.Configuration;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.design.widget.TabLayout;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import com.aquamorph.frcmanager.R;
import com.aquamorph.frcmanager.adapters.SectionsPagerAdapter;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;

public class MainActivity extends AppCompatActivity implements OnSharedPreferenceChangeListener {


	private String TAG = "MainActivity";
	private SectionsPagerAdapter mSectionsPagerAdapter;
	private ViewPager mViewPager;
	public String teamNumber, eventName;
	public Boolean paidUser = false;
	public String year = "2015";
	public String teamRank;
	Toolbar toolbar;
	SharedPreferences prefs;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);

		prefs = PreferenceManager.getDefaultSharedPreferences(this);
		prefs.registerOnSharedPreferenceChangeListener(MainActivity.this);
		teamNumber = prefs.getString("teamNumber", "");
		eventName = prefs.getString("eventShortName", "");
		year = prefs.getString("teamNumber", "2015");
		teamRank = prefs.getString("teamRank", "");

		if (teamNumber.equals("")) openSetup();


		listener();
		theme(this);
	}

	public static void theme(Activity activity) {
		SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(activity);
		switch (prefs.getString("theme", "")) {
			case "amoled":
				activity.setTheme(R.style.OMOLEDTheme);
				break;
			case "dark":
				activity.setTheme(R.style.DarkTheme);
				break;
			default:
				activity.setTheme(R.style.LightTheme);
				break;
		}
	}

	@Override
	public void onConfigurationChanged(Configuration newConfig) {
		super.onConfigurationChanged(newConfig);
		setContentView(R.layout.activity_main);
		listener();
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		// Inflate the menu; this adds items to the action bar if it is present.
		getMenuInflater().inflate(R.menu.menu_main, menu);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		// Handle action bar item clicks here. The action bar will
		// automatically handle clicks on the Home/Up button, so long
		// as you specify a parent activity in AndroidManifest.xml.
		int id = item.getItemId();

		//noinspection SimplifiableIfStatement
		if (id == R.id.action_settings) {
			openSettings();
			return true;
		}
		return super.onOptionsItemSelected(item);
	}

	private void openSetup() {
		Intent intent = new Intent(this, Setup.class);
		startActivity(intent);
	}

	public void openSettings() {
		Intent intent = new Intent(this, Settings.class);
		startActivity(intent);
	}

	public String getTeamNumber() {
		return "frc" + teamNumber;
	}

	public String getSubTitle() {
		if(teamRank.equals("")) {
			return String.format("%s (%s)", eventName, teamNumber);
		} else {
			return String.format("%s (%s) Rank #%s", eventName, teamNumber, teamRank);
		}
	}

	@Override
	public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
		teamNumber = sharedPreferences.getString("teamNumber", "0000");
		eventName = sharedPreferences.getString("eventShortName", "North Carolina");
		teamRank = sharedPreferences.getString("teamRank", "");
		if (getSupportActionBar() != null) {
			getSupportActionBar().setSubtitle(getSubTitle());
		}
		if (key.equals("theme")) {
			this.recreate();
		}
	}

	private void listener() {
		toolbar = (Toolbar) findViewById(R.id.toolbar);
		setSupportActionBar(toolbar);
		toolbar.setSubtitle(getSubTitle());

		// Create the adapter that will return a fragment for each of the three
		// primary sections of the activity.
		mSectionsPagerAdapter = new SectionsPagerAdapter(getSupportFragmentManager());

		// Set up the ViewPager with the sections adapter.
		mViewPager = (ViewPager) findViewById(R.id.container);
		mViewPager.setOffscreenPageLimit(mSectionsPagerAdapter.getCount());
		mViewPager.setAdapter(mSectionsPagerAdapter);

		TabLayout tabLayout = (TabLayout) findViewById(R.id.tabs);
		tabLayout.setupWithViewPager(mViewPager);
		tabLayout.setTabMode(TabLayout.MODE_SCROLLABLE);

		//Load ads
		AdView mAdView = (AdView) findViewById(R.id.adView);
		if (paidUser) {
			mAdView.setVisibility(View.GONE);
		} else {
			AdRequest adRequest = new AdRequest.Builder()
					.addTestDevice(getResources().getString(R.string.nexus_5_test_id))
					.addTestDevice(getResources().getString(R.string.moto_g_test_id))
					.addTestDevice(getResources().getString(R.string.neux_6p_test_id))
					.build();
			mAdView.loadAd(adRequest);
		}
	}
}

File: app/src/main/java/com/aquamorph/frcmanager/activities/MainActivity.kt
package com.aquamorph.frcmanager.activities

import android.app.Activity
import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.content.res.Configuration
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.design.widget.TabLayout
import android.support.v4.view.ViewPager
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuItem

import com.aquamorph.frcmanager.R
import com.aquamorph.frcmanager.adapters.SectionsPagerAdapter
import com.aquamorph.frcmanager.network.DataLoader
import com.aquamorph.frcmanager.utils.Constants
import com.aquamorph.frcmanager.utils.Logging

/**
 * Default activity of the app.
 *
 * @author Christian Colglazier
 * @version 3/29/2016
 */
class MainActivity : AppCompatActivity(), OnSharedPreferenceChangeListener {
    private var eventName: String? = null
    private var teamRank: String? = null
    private var teamRecord: String? = null
    lateinit var dataLoader: DataLoader
    private lateinit var mViewPager: ViewPager
    private var tabLayout: TabLayout? = null

    /**
     * getTeamNumber() returns the string tag for Blue Alliance requests of the current
     * team number.
     *
     * @return team number string
     */
    val teamNumber: String
        get() = "frc" + DataLoader.teamNumber

    /**
     * getAppTitle() returns the text for the app.
     * @return title for the app
     */
    val appTitle: String
        get() = String.format("%s - %s", DataLoader.teamNumber,
                shorten(eventName, Constants.MAX_EVENT_TITLE_LENGTH))

    /**
     * getAppSubTitle() returns the subtitle string for the toolbar with the event name,
     * team number and current rank if available.
     *
     * @return subtitle string
     */
    private val appSubTitle: String
        get() = if (teamRank == "") {
            ""
        } else {
            String.format("Rank #%s %s", teamRank, teamRecord)
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val prefs = PreferenceManager.getDefaultSharedPreferences(this)
        prefs.registerOnSharedPreferenceChangeListener(this@MainActivity)
        DataLoader.teamNumber = prefs.getString("teamNumber", "")
        eventName = prefs.getString("eventShortName", "")
        teamRank = prefs.getString("teamRank", "")
        teamRecord = prefs.getString("teamRecord", "")
        DataLoader.eventKey = prefs.getString("eventKey", "")


        if (DataLoader.teamNumber == "") openSetup()

        listener()
        theme(this)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        setContentView(R.layout.activity_main)
        Logging.info(this, "onConfigurationChanged", 2)
        listener()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_settings -> openSettings()
            R.id.refresh_all -> mSectionsPagerAdapter!!.refreshAll(false)
            else -> mSectionsPagerAdapter!!.refreshAll(false)
        }
        return super.onOptionsItemSelected(item)
    }

    /**
     * openSetup() launches the set up activity.
     */
    private fun openSetup() {
        val intent = Intent(this, Setup::class.java)
        startActivity(intent)
    }

    /**
     * openSettings() launches the settings activity.
     */
    fun openSettings() {
        val intent = Intent(this, Settings::class.java)
        startActivity(intent)
    }

    /**
     * shorten() returns a shortened string with ... at the end.
     *
     * @param text   to be shortened
     * @param amount length to shorten
     * @return shorten string with ... at the end
     */
    fun shorten(text: String?, amount: Int): String {
        return if (text!!.length > amount) {
            text.substring(0, amount) + "..."
        } else {
            text
        }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        eventName = sharedPreferences.getString("eventShortName", "North Carolina")
        teamRank = sharedPreferences.getString("teamRank", "")
        teamRecord = sharedPreferences.getString("teamRecord", "")
        when (key) {
            "eventKey" -> DataLoader.eventKey = sharedPreferences.getString("eventKey", "")
            "teamNumber" -> DataLoader.teamNumber = sharedPreferences.getString("teamNumber", "0000")
            else -> {
            }
        }

        if (supportActionBar != null) {
            supportActionBar!!.title = appTitle
            supportActionBar!!.subtitle = appSubTitle
        }
        if (key == "theme") {
            this.recreate()
        }
    }

    /**
     * listener() connects layout to view.
     */
    private fun listener() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar!!.title = appTitle
        supportActionBar!!.subtitle = appSubTitle

        // Set up the ViewPager with the sections adapter.
        mViewPager = findViewById(R.id.container)
        tabLayout = findViewById(R.id.tabs)
        if (tabLayout != null) {
            tabLayout!!.setupWithViewPager(mViewPager)
            tabLayout!!.tabMode = TabLayout.MODE_SCROLLABLE
        }
        mSectionsPagerAdapter = SectionsPagerAdapter(supportFragmentManager,
                mViewPager, tabLayout!!, this)
        dataLoader = DataLoader(this, mSectionsPagerAdapter!!)
        mViewPager.offscreenPageLimit = mSectionsPagerAdapter!!.count
        mViewPager.adapter = mSectionsPagerAdapter
        refreshData(false)
    }

    companion object {

        private var mSectionsPagerAdapter: SectionsPagerAdapter? = null

        /**
         * theme() sets the app theme based on user selection from settings.
         *
         * @param activity
         */
        fun theme(activity: Activity) {
            val prefs = PreferenceManager.getDefaultSharedPreferences(activity)
            when (prefs.getString("theme", "")) {
                "amoled" -> activity.setTheme(R.style.OMOLEDTheme)
                "dark" -> activity.setTheme(R.style.DarkTheme)
                else -> activity.setTheme(R.style.LightTheme)
            }
        }

        fun refreshData(force: Boolean) {
            mSectionsPagerAdapter!!.refrestData(force)
        }

        fun refresh() {
            mSectionsPagerAdapter!!.refreshAll(false)
        }

        fun refresh(force: Boolean?) {
            mSectionsPagerAdapter!!.refreshAll(force!!)
        }
    }
}
